build/pulsar_ci_tool.sh

```diff
@@ -46,6 +47,123 @@ function ci_dependency_check() {
   _ci_mvn -Pmain,skip-all,skipDocker,owasp-dependency-check initialize verify -pl '!pulsar-client-tools-test' "$@"
 }
 
+# installs a tool executable if it's not found on the PATH
+function ci_install_tool() {
+  local tool_executable=$1
+  local tool_package=${2:-$1}
+  if ! command -v $tool_executable &>/dev/null; then
+    echo "::group::Installing ${tool_package}"
+    sudo apt-get -y install ${tool_package} >/dev/null
+    echo '::endgroup::'
+  fi
+}
+
+# outputs the given message to stderr and exits the shell script
+function fail() {
+  echo "$*" >&2
+  exit 1
+}
+
+# function to retry a given commend 3 times with a backoff of 10 seconds in between
+function ci_retry() {
+  local n=1
+  local max=3
+  local delay=10
+  while true; do
+    "$@" && break || {
+      if [[ $n -lt $max ]]; then
+        ((n++))
+        echo "::warning::Command failed. Attempt $n/$max:"
+        sleep $delay
+      else
+        fail "::error::The command has failed after $n attempts."
+      fi
+    }
+  done
+}
+
+# saves a given image (1st parameter) to the GitHub Actions cache with the given key (2nd parameter)
+function ci_docker_save_image_to_github_actions_cache() {
+  local image=$1
+  local cachekey=$2
+  ci_install_tool pv
+  echo "::group::Saving docker image ${image} with key ${cachekey} in GitHub Actions Cache"
+  ci_retry bash -c "docker save ${image} | zstd | pv -ft -i 5 | pv -Wbaf -i 5 | curl -s -H 'Content-Type: application/octet-stream' -X PUT --data-binary @- http://localhost:12321/${cachekey}"
+  echo "::endgroup::"
+}
+
+# loads a docker image from the GitHub Actions cache with the given key (1st parameter)
+function ci_docker_load_image_from_github_actions_cache() {
+  local cachekey=$1
+  ci_install_tool pv
+  echo "::group::Loading docker image from key ${cachekey} in GitHub Actions Cache"
+  ci_retry bash -c "curl -s http://localhost:12321/${cachekey} | pv -batf -i 5 | unzstd | docker load"
```

Comment:
is this port documented somewhere ?

Reply:
> IIUC now the pulsarbot is not needed anymore and you have to enter the workflow run and re-run the single job, is it ?

It's needed. This doesn't merge all the workflows to one. It's the majority of the workflows that are merged. 
The bot is needed so that non-committers can restart failed jobs. It might require a tweak to pulsarbot so that it uses the GitHub Actions API to start just the failed jobs in a workflow and not the complete workflow and all jobs in it.

Reply:
Here's the documentation: https://github.com/cirruslabs/http-cache-action#port 
The port 12321 is used for the http-cache-action which is used to access GitHub Actions cache with a HTTP interface.
The GitHub Actions cache doesn't currently have a CLI for storing arbitrary binary streams. That feature is needed in this case.

Reply:
@yaalsn Each user has 10 GB for pull requests in GitHub Actions Cache. That's part of the security model. You cannot make updates to the upstream repository's cache where you open a PR. It's hierarchical internally. Updates are only made in the PR owner's cache and impact it's quota.

Reply:
the https://docs.google.com/document/d/1FNEWD3COdnNGMiryO9qBUW_83qtzAhqjDI5wwmPD-YE/edit# doc also explains this.
> GitHub Actions Cache is restricted to the scope of the forked repository in pull requests to prevent security issues around cache poisoning. The additional benefit is that the total capacity of caches is beyond 10GB since the cache is not shared between forks. Each fork has a quota of 10GB. The GitHub Actions Cache can be used as a general purpose file storage to share the build artifacts as files between the stages.

Reply:
I started a thread in GitHub Community support forum to clarify this: https://github.community/t/how-does-the-github-actions-cache-size-limit-work-for-pull-requests/240546

Reply:
In apache/pulsar builds, there will be the 10GB limit for GitHub Actions cache. The cache is a LRU cache based on the [description in actions/cache README](https://github.com/actions/cache#cache-limits):
> A repository can have up to 10GB of caches. **Once the 10GB limit is reached, older caches will be evicted based on when the cache was last accessed.** Caches that are not accessed within the last week will also be evicted.

This behavior is optimal. Each build will temporarily create about 2.5GB of cache entries. Since the needed cache entries are accessed frequently, this will ensure that important caches (the maven dependencies caches) aren't likely to be dropped. maven dependencies caches are updated with a scheduled job once a day: https://github.com/apache/pulsar/blob/master/.github/workflows/ci-maven-cache-update.yaml

The 10GB limit could cause issues if there are too many concurrent build jobs for apache/pulsar. That problem can be addressed by falling back to building the binaries if there's a cache miss. That solution could be added later if problems occur. GitHub actions has also concurrency limits which are added in this PR. A more recent build will cancel a build that started earlier for a given branch. In the context of apache/pulsar, the builds will be pushes to master and branch-* branches. It would be a very rare case where there's more than 4 branches that are built at the same time. It's possible, but if that happens, the build can be improved to handle cache misses.

Reply:
I have also asked the question in https://github.com/actions/cache/discussions/776 .

Reply:
I'm writing a command line client for GitHub Actions cache so that it would integrate with the refactored build. The current work-in-progress is at https://github.com/lhotari/gh-actions-artifact-client . It uses the official GitHub Actions javascript npm modules for interacting with GitHub Actions Artifact backend.

Reply:
GitHub Actions Artifact has a retention period option, the artifacts could be cleaned after all the workflow is finished.

Reply:
> GitHub Actions Artifact has a retention period option, the artifacts could be cleaned after all the workflow is finished.

Yes. I think it's sufficient to let the retention period to take care of the clean up. I now have https://github.com/lhotari/gh-actions-artifact-client working and I tested it with a 600MB file. The performance is pretty good. About 11MB/s for writing and over 23 MB/s for reading (with limited testing). There's a chance to improve the performance further by adding some concurrency.
I'll start working on the changes to integrate it to the Pulsar CI refactoring.